pymc_experimental/utils/model_fgraph.py

```diff
@@ -0,0 +1,174 @@
+
+
+class ModelVar(Op):
+    """A dummy Op that describes the purpose of a Model variable and contains
+    meta-information as additional inputs (value and dims).
+    """
+
+    def make_node(self, rv, value=None, dims: Optional[Sequence[str]] = None):
+        assert isinstance(rv, TensorVariable)
+
+        if dims is not None:
+            dims = [pytensor.as_symbolic(dim) for dim in dims]
+            assert all(isinstance(dim.type, StringType) for dim in dims)
+            assert len(dims) == rv.type.ndim
+        else:
+            dims = ()
+
+        if value is not None:
+            assert isinstance(value, TensorVariable)
+            assert rv.type.in_same_class(value.type)
+            return Apply(self, [rv, value, *dims], [rv.type()])
+        else:
+            return Apply(self, [rv, *dims], [rv.type()])
+
+    def infer_shape(self, fgraph, node, inputs_shape):
+        return inputs_shape[0]
+
+    def do_constant_folding(self, fgraph, node):
+        return False
+
+    def perform(self, *args, **kwargs):
+        raise RuntimeError("ValuedRVs should never be evaluated!")
+
+
+class FreeRV(ModelVar):
+    pass
+
+
+class ObservedRV(ModelVar):
+    pass
+
+
+class Potential(ModelVar):
+    pass
+
+
+class Deterministic(ModelVar):
+    pass
+
+
+free_rv = FreeRV()
+observed_rv = ObservedRV()
+potential = Potential()
+deterministic = Deterministic()
+
+
+def toposort_replace(fgraph: FunctionGraph, replacements) -> None:
+    """Replace multiple variables in topological order."""
+    toposort = fgraph.toposort()
+    sorted_replacements = sorted(replacements, key=lambda pair: toposort.index(pair[0].owner))
+    fgraph.replace_all(tuple(sorted_replacements), import_missing=True)
+
+
+def fgraph_from_model(model: Model) -> FunctionGraph:
+
+    # Collect PyTensor variables
+    rvs_to_values = model.rvs_to_values
+    rvs = list(rvs_to_values.keys())
+    values = list(rvs_to_values.values())
+    free_rvs = model.free_RVs
+    deterministics = model.deterministics
+    potentials = model.potentials
+
+    # Collect PyMC meta-info
+    vars_to_dims = model.named_vars_to_dims
+    coords = model.coords
+
+    # TODO: Do something with these
+    dim_lengths = model.dim_lengths
+    rvs_to_transforms = model.rvs_to_transforms
+
+    # Not supported yet
+    if any(v is not None for v in model.rvs_to_total_sizes.values()):
+        raise NotImplementedError("Cannot convert models with total_sizes")
+    if any(v is not None for v in model.rvs_to_initial_values.values()):
+        raise NotImplementedError("Cannot convert models with non-default initial_values")
+
+    # We start the `dict` with mappings from the value variables to themselves,
+    # to prevent them from being cloned.
+    memo = {v: v for v in values}
+
+    fgraph = FunctionGraph(
+        outputs=rvs + potentials + deterministics,
+        clone=True,
+        memo=memo,
+        copy_orphans=False,
+        copy_inputs=False,
```

Comment:
Does this mean that each `RandomGenerator` will get its own copy of the shared random stream? Or will that underlying memory still be shared? If yes, that isn't what I think that we want. I think that we want the new variables to use their own copied versions of the random streams or generators. By copies I mean that they should be two different generator objects that have the exact same initial random state. That way, they would yield the same sequence of random numbers, but getting random numbers from the new model would not alter the state of the old model.

Reply:
I don't know what happens to shared variables, here I was thinking of not copying the non-shared value variables. Will have to take a look. I think we should copy them, just so that two compiled random functions don't interfere with each other.

About the initial state, that doesn't matter. Whenever we compile a function with `pytensorf.compile_pymc` we override the RNG values based on `random_seed` anyway. The current global values are never taken into consideration (and I think they shouldn't as it creates a very cumbersome internal state to our PyMC models)

Reply:
Actually we probably want to reuse shared variables other than RNGs (eg mutableData), so maybe I will do a replace rngs after cloning but otherwise reuse the same shared variables

Reply:
Replaced RNGs now